core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright 2020 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Segment of a PMMR.
+
+use crate::core::hash::Hash;
+use crate::core::pmmr::{self, Backend, ReadablePMMR, ReadonlyPMMR};
+use crate::ser::{Error, PMMRIndexHashable, PMMRable, Readable, Reader, Writeable, Writer};
+use croaring::Bitmap;
+use std::cmp::min;
+use std::fmt::{self, Debug};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// Error related to segment creation or validation
+pub enum SegmentError {
+	/// An expected leaf was missing
+	MissingLeaf(u64),
+	/// An expected hash was missing
+	MissingHash(u64),
+	/// The segment does not exist
+	NonExistent,
+	/// The segment is fully pruned
+	Pruned,
+	/// Mismatch between expected and actual root hash
+	Mismatch,
+}
+
+impl fmt::Display for SegmentError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			SegmentError::MissingLeaf(idx) => write!(f, "Missing leaf at pos {}", idx),
+			SegmentError::MissingHash(idx) => write!(f, "Missing hash at pos {}", idx),
+			SegmentError::NonExistent => write!(f, "Segment does not exist"),
+			SegmentError::Pruned => write!(f, "Segment is fully pruned"),
+			SegmentError::Mismatch => write!(f, "Root hash mismatch"),
+		}
+	}
+}
+
+/// Tuple that defines a segment of a given PMMR
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct SegmentIdentifier {
+	/// 2-Log of the size of a segment
+	pub log_size: u8,
```

Comment:
I think `log_size` is actually the same as our existing `height` terminology -

https://github.com/mimblewimble/grin/blob/cea546ceb8b5be0fc4462489e8a4761bc395fb18/core/src/core/pmmr/pmmr.rs#L540-L548

Would it make sense to use `height` here to avoid introducing the new `log_size` terminology?